### PR TITLE
Implemented question VC

### PR DESCRIPTION
**Pull request by zhiachong**
TODO:
Complete the questions, and complete the animation.
Refactor code

profile

### CODE REVIEW

File: ios-trivia-game/API/FirebaseClient.swift [hunk 1]
```diff
@@ -265,6 +316,35 @@ class FirebaseClient {
...
+        self.getQuestionsBy(categoryId: "\(categoryId)", complete: {(snapshot) in
+            var ids : [Int] = []
+            print ("val \(snapshot.value)")
+            
+            if let questions = snapshot.value as? NSDictionary {
+                let allQuestions = questions.allValues
+                for questionData in allQuestions {
+                    if let questionDict = questionData as? NSDictionary {
+                        let triviaQuestion = TriviaQuestion(dictionary: questionDict)
+                        ids.append(triviaQuestion.id!)
```
Comment (saviotsui) [inline ios-trivia-game/API/FirebaseClient.swift:387]: Why store just the Ids?
Since you've already grabbed the questions already you might be able to save network calls back to Firebase by passing the question set around

File: ios-trivia-game/Constants.swift [hunk 2]
```diff
@@ -16,12 +16,17 @@ class Constants {
...
+    static let SUGGESTION_TABLE_NAME = "suggestions"
+
     
     // For segues
     static let LOGIN_MODAL_SEGUE = "com.iostriviagame.loginSegue"
+    static let QUESTION_TO_ANSWER_SEGUE = "com.iostriviagame.questiontoanswersegue"
     static let ANSWER_TO_RESULTS_SEGUE = "com.iostriviagame.answertoresultssegue"
     static let GAME_OPTIONS_TO_COUNTDOWN_SEGUE = "com.iostriviagame.gameoptionstocountdownsegue"
     
+
```
Comment (saviotsui) [inline ios-trivia-game/Constants.swift]: nit: extra line

File: ios-trivia-game/Constants.swift [hunk 3]
```diff
@@ -16,12 +16,17 @@ class Constants {
     static let ANSWER_TABLE_NAME = "answers"
     static let SCORED_ANSWER_TABLE_NAME = "scored_answers"
     static let INVITE_TABLE_NAME = "invites"
+    static let CATEGORY_TABLE_NAME = "categories"
+    static let SUGGESTION_TABLE_NAME = "suggestions"
+
```
Comment (saviotsui) [inline ios-trivia-game/Constants.swift:21]: nit: extra line

File: ios-trivia-game/API/FirebaseClient.swift (hunk 1 above)
Comment (saviotsui) [reply ios-trivia-game/API/FirebaseClient.swift:387]: We could probably do the same with the GameRoom entity once we are in a game.

File: ios-trivia-game/Models/Answer.swift [hunk 4]
```diff
@@ -31,6 +31,16 @@ class Answer: NSObject {
         self.timestamp = String(describing: NSDate())
     }
     
+    init(answerText: String) {
```
Comment (saviotsui) [inline ios-trivia-game/Models/Answer.swift]: why do you need this initializer?
since you need to populate these values, why not use the initializer above?

Remove this initializer and use the one above

File: ios-trivia-game/Models/TriviaQuestion.swift [hunk 5]
```diff
@@ -33,16 +33,16 @@ class TriviaQuestion: NSObject {
     var answer: String?
     var question: String?
     var value: Int?
-    // var category: TriviaCategory?
-    var roomId: Int?
+    var roomId: String?
```
Comment (saviotsui) [inline ios-trivia-game/Models/TriviaQuestion.swift]: I think we can safely remove roomId from this entity

File: ios-trivia-game/ViewControllers/AnswerViewController.swift [hunk 6]
```diff
@@ -43,6 +45,13 @@ class AnswerViewController: UIViewController {
                 Logger.instance.log(logLevel: .info, message: "Adding \(answer.getJson())")
             }
             
+            // add the correct answer
+            let correctAnswer = self.question?.answer ?? ""
+            if (correctAnswer != "") {
+                let answer = Answer(answerText: correctAnswer)
```
Comment (saviotsui) [inline ios-trivia-game/ViewControllers/AnswerViewController.swift]: I think you need to rethink the code here.

In the questionviewcontroller, before you segue to AnswerViewController, you should register the correct answer to the database.

You already have the roomid and the question id as well as the correct answer text.

File: ios-trivia-game/ViewControllers/GameOptionsViewController.swift [hunk 7]
```diff
@@ -78,18 +84,32 @@ class GameOptionsViewController: UIViewController {
 
     // Make an api call to create a game room, and then go to CountdownTimerViewController if it succeeds.
     @IBAction func onStartGameClicked(_ sender: Any) {
-        let newGame = GameRoom(id: FirebaseClient.instance.createGameRoomId().key, name: self.nameOfGameroom, currentNumPlayers: 0, maxNumPlayers: self.numOfPlayers, state: GameRoom.State.idle, isPublic: self.isPublic, currentQuestion: 0, maxNumQuestions: self.numOfQuestions)
-
-        FirebaseClient.instance.createGame(gameRoom: newGame.getJson() as NSDictionary, complete: {_, ref in
-            let roomId = ref.key
+        FirebaseClient.instance.getRandomQuestions(categoryId: category!, maxNumOfQuestions: numOfQuestions!, complete: {(questionList) in
+            let newGame = ["max_num_of_questions" : self.numOfQuestions ?? 0,
```
Comment (saviotsui) [inline ios-trivia-game/ViewControllers/GameOptionsViewController.swift]: why did you wipe out line 81?
You're duplicating the default values in the initializer for creating a game

revert this

File: ios-trivia-game/ViewControllers/GameOptionsViewController.swift [hunk 8]
```diff
@@ -78,18 +84,32 @@ class GameOptionsViewController: UIViewController {
...
-            let storyboard = UIStoryboard(name: "Main", bundle: nil)
-            let destination = storyboard.instantiateViewController(withIdentifier: Constants.COUNTDOWN_NAVIGATION_VIEW_CONTROLLER)
-            let countdownNavigationController = destination as! UINavigationController
-            let countdownGameViewController = countdownNavigationController.topViewController as! CountdownGameViewController
-            countdownGameViewController.roomId = roomId
-            self.present(destination, animated: true, completion: nil)
+            FirebaseClient.instance.createGame(gameRoom: newGame as NSDictionary, complete: {_, ref in
+                let roomId = ref.key
+                
+                print("Created game room with roomID: \(roomId)")
```
Comment (saviotsui) [inline ios-trivia-game/ViewControllers/GameOptionsViewController.swift:98]: revert this

File: ios-trivia-game/ViewControllers/GameOptionsViewController.swift [hunk 9]
```diff
@@ -101,6 +121,26 @@ class GameOptionsViewController: UIViewController {
...
     }
     */
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        // do some customization
+        
+        let gameRoom = GameRoom()
+        gameRoom.current_question = 0
+        // this needs to be retrieved from create game
+        gameRoom.is_public = true
+        gameRoom.max_num_of_people = 10
```
Comment (saviotsui) [inline ios-trivia-game/ViewControllers/GameOptionsViewController.swift]: revert this

File: ios-trivia-game/ViewControllers/QuestionViewController.swift [hunk 10]
```diff
@@ -11,32 +11,68 @@ import UIKit
...
+            let curQuestionIndex = gameRoom.current_question
+            let questions = gameRoom.questions
+            let maxQuestions = gameRoom.max_num_of_questions
+            
+            self.questionNumber.text = "Question \(curQuestionIndex! + 1)/\(maxQuestions)"
+            
+            
+            if (curQuestionIndex! >= questions.count) {
+                // prevent index-out-of-bound error
+                return
```
Comment (saviotsui) [inline ios-trivia-game/ViewControllers/QuestionViewController.swift:53]: why are you returning here? you should transition to the finalscoreviewcontroller

File: ios-trivia-game/ViewControllers/QuestionViewController.swift [hunk 11]
```diff
@@ -11,32 +11,68 @@ import UIKit
...
+            
+            FirebaseClient.instance.getQuestionBy(questionId: curQuestionId, complete: { (snapshot) in
+                if let questionDict = snapshot.value as? NSDictionary {
+                    let question = questionDict.value(forKey: "\(curQuestionId)") as? NSDictionary
+                    self.question = TriviaQuestion(dictionary: question!)
+                }
+                
+                self.questionTitle.text = self.question?.question
+                
+                FirebaseClient.instance.getSuggestionsBy(questionId: curQuestionId, complete: {(snapshot) in
```
Comment (saviotsui) [inline ios-trivia-game/ViewControllers/QuestionViewController.swift]: as discussed, please remove the suggestions code until we complete our basic game

File: ios-trivia-game/ViewControllers/QuestionViewController.swift [hunk 12]
```diff
@@ -50,12 +86,78 @@ class QuestionViewController: UIViewController {
...
+    @IBAction func quitGame(_ sender: Any) {
+        countdownTimer.invalidate()
+        Utilities.quitGame(controller: self)
+    }
 
     @IBAction func onAnswerSubmit(_ sender: UIButton) {
         // Temporary to demonstrate AnswerViewController
         // @Nari or @Zhia: Feel free to remove the hardcoded values when you get to this
+        let answerText = self.answerField.text
+
```
Comment (saviotsui) [inline ios-trivia-game/ViewControllers/QuestionViewController.swift:99]: call countdownTimer.invalidate()

File: ios-trivia-game/ViewControllers/ResultsViewController.swift [hunk 13]
```diff
@@ -45,12 +45,15 @@ class ResultsViewController: UIViewController {
...
                 
                 self.resultsTitleLabel.text = "Results: Round \(self.gameRoom!.current_question!) of \(self.gameRoom!.max_num_of_questions)"
                 
-                if (self.gameRoom?.current_question == self.gameRoom?.max_num_of_questions) {
-                    self.nextQuestionButton.isHidden = true
+                let curQues = self.gameRoom?.current_question
+                let maxQuestions = self.gameRoom?.max_num_of_questions
+                if (curQues! >= maxQuestions!) {
+                    // needs to be completed
+                    //self.nextQuestionButton.isHidden = true
```
Comment (saviotsui) [inline ios-trivia-game/ViewControllers/ResultsViewController.swift:56]: you need to hide the next question button if you've reached max number of questions no?

File: ios-trivia-game/ViewControllers/ResultsViewController.swift [hunk 14]
```diff
@@ -94,7 +97,9 @@ class ResultsViewController: UIViewController {
             let storyboard = UIStoryboard(name: "Main", bundle: nil)
             let destinationIdentifier: String
             let destination: UIViewController
-            if (self.gameRoom?.current_question == self.gameRoom?.max_num_of_questions) {
+            let curQuestion = self.gameRoom?.current_question
+            let maxQuestions = self.gameRoom?.max_num_of_questions
```
Comment (saviotsui) [inline ios-trivia-game/ViewControllers/ResultsViewController.swift:101]: why did you change this? 97 functionally does the same thing

**Review by saviotsui — CHANGES_REQUESTED** [review]

File: ios-trivia-game/API/FirebaseClient.swift (hunk 1 above)
Comment (zhiachong) [reply ios-trivia-game/API/FirebaseClient.swift:387]: Other users joining the room wouldn't be able to participate in the same game with the same set of questions if we don't store the question IDs. 

We can return a set of questions instead of question IDs, but we would still need to store the question Ids.

File: ios-trivia-game/ViewControllers/AnswerViewController.swift (hunk 6 above)
Comment (zhiachong) [reply ios-trivia-game/ViewControllers/AnswerViewController.swift]: It's not necessary to store the correct answer to the database because every other user would send a request to store the correct answer in the database. The primary goal here is that the correct answer would be used on client-side, rather than being retrieved from the server-side for each user.

If the correct answer is stored in DB, then the scenario is we would have N number of correct answers, where N is the number of players because everyone would've submitted a correct answer from the question + their own answer. 

We can talk about this further.

File: ios-trivia-game/ViewControllers/GameOptionsViewController.swift (hunk 7 above)
Comment (zhiachong) [reply ios-trivia-game/ViewControllers/GameOptionsViewController.swift]: Hmm to me it looks like both are doing the same things, except that in line 87, `newGame` is an associative array, whereas line 81 is using a class? 

Sorry I'm not seeing where the duplication is happening :p

File: ios-trivia-game/ViewControllers/GameOptionsViewController.swift (hunk 8 above)
Comment (zhiachong) [reply ios-trivia-game/ViewControllers/GameOptionsViewController.swift:98]: My comment above.

File: ios-trivia-game/ViewControllers/GameOptionsViewController.swift (hunk 9 above)
Comment (zhiachong) [reply ios-trivia-game/ViewControllers/GameOptionsViewController.swift]: I'm not 100% sure when this is used because I didn't work on it. I think I included this when I tried to resolve conflicts. 

If removing this section is fine, then I'll go ahead and remove it.

File: ios-trivia-game/ViewControllers/HomeViewController.swift [hunk 15]
```diff
@@ -70,7 +70,7 @@ class HomeViewController: UIViewController {
         FirebaseClient.instance.getGameRooms(complete: { (dictionary) in
             for (roomId, roomInfo) in dictionary {
                 let gameRoom = GameRoom(dictionary: roomInfo as! NSDictionary)
-                if gameRoom.state != GameRoom.State.end {
+                if gameRoom.state == GameRoom.State.idle {
```
Comment (zhiachong) [inline ios-trivia-game/ViewControllers/HomeViewController.swift]: @plushysmurf3 As talked about, I'm checking games in idle state here to make sure the home view controller only shows games that a user can join.

File: ios-trivia-game/ViewControllers/QuestionViewController.swift (hunk 10 above)
Comment (zhiachong) [reply ios-trivia-game/ViewControllers/QuestionViewController.swift:53]: So firebase does something odd here. When I increment the current_question count in the database, somehow even w/o invoking the method, the `curQuestionIndex` was updated. I kept running into an index-out-of-bounds error, and thus I'm short-circuiting the code. If I had transitioned to final score, the flow would be incorrect because the user would transition to the final score right after question. 

I didn't spend too much time debugging it and found the path of least resistance :/

File: ios-trivia-game/ViewControllers/QuestionViewController.swift (hunk 12 above)
Comment (zhiachong) [reply ios-trivia-game/ViewControllers/QuestionViewController.swift:99]: I don't think calling invalidate() here is desired because that means the user would be brought to the answer view right away. The flow in my mind is: 

Submit your answer -> Wait until time runs out -> See the list of answers from everyone -> Pick your correct answer

If we don't, then step 3 in the list of answers would only contain answers from whoever submitted first.

File: ios-trivia-game/ViewControllers/ResultsViewController.swift (hunk 13 above)
Comment (zhiachong) [reply ios-trivia-game/ViewControllers/ResultsViewController.swift:56]: There's no next question button from what I can tell. It's not bounded to any label/button right now. Which is why I didnt' set it.

File: ios-trivia-game/ViewControllers/ResultsViewController.swift (hunk 14 above)
Comment (zhiachong) [reply ios-trivia-game/ViewControllers/ResultsViewController.swift:101]: Where's 97? What's that?

**Comment by zhiachong** [comment]
Argh, main storyboard conflict again. I hate github :/ 

@plushysmurf3 good comments and suggestions! I'll fix the suggestions portion, and I left my thoughts on some of the things you pointed out. Let me know if we need to discuss them further.

File: ios-trivia-game/ViewControllers/QuestionViewController.swift (hunk 12 above)
Comment (saviotsui) [reply ios-trivia-game/ViewControllers/QuestionViewController.swift:99]: No. Invalidate() kills the timer. you need to killer the timer.

Comment (zhiachong) [reply ios-trivia-game/ViewControllers/QuestionViewController.swift:99]: Why do we need to kill the timer? Shouldn't the user submit an answer, and then wait until the 60 seconds run-out, and then move to the answer page? Slightly confused here.

File: ios-trivia-game/API/FirebaseClient.swift (hunk 1 above)
Comment (zhiachong) [reply ios-trivia-game/API/FirebaseClient.swift:387]: Optional: 

Refactor if possible to pass the set of questions around, instead of making multiple network calls.

File: ios-trivia-game/Models/GameRoom.swift [hunk 16]
```diff
@@ -22,26 +22,30 @@ class GameRoom: NSObject {
...
+    init(dictionary: NSDictionary = NSDictionary()) {
         self.id = dictionary["id"] as? String ?? "0"
         self.name = dictionary["name"] as? String ?? "-- Unnamed Room --"
         self.current_num_players = dictionary["current_num_players"] as? Int ?? 0
         self.max_num_of_people = dictionary["max_num_of_people"] as? Int ?? MAX_NUMBER_OF_PEOPLE
         self.state = State(rawValue: dictionary["state"] as! Int)!
         self.is_public = dictionary["is_public"] as? Bool ?? true
         self.current_question = dictionary["current_question"] as? Int ?? 0
+        self.category = dictionary["category"] as? Int ?? 0
         self.max_num_of_questions = dictionary["max_num_of_questions"] as? Int ?? MAX_NUMBER_OF_QUESTIONS
```
Comment (saviotsui) [inline ios-trivia-game/Models/GameRoom.swift:43]: is this redundant now since we have self.questions?

**Review by saviotsui — APPROVED** [review]
